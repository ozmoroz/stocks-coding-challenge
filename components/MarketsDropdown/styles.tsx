import styled from 'styled-components';
import { Button } from 'react-bootstrap';

export const Dropdown = styled.div`
  position: relative;
  display: inline-block;
  width: 100%;
`;

export const DropdownButton = styled(Button)`
  border-color: rgb(206, 212, 218);
`;

export const DropdownContent = styled.ul`
  list-style: none;
  position: absolute;
  background-color: #fff;
  min-width: 160px;
  box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
  z-index: 100;
  padding: 0;
`;

interface IDropdownItem {
  /** Is the current element highlighted? */
  isHighlighted?: boolean;
}

/** Highlight the currently selected list item */
export const DropdownItem = styled.li<IDropdownItem>`
  padding: 0 0.5rem;
  background-color: ${(props) => (props.isHighlighted ? '#bde4ff' : '#fff')};
`;
