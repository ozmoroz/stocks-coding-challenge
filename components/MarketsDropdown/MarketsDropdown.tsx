import { useState } from 'react';
import { useCombobox } from 'downshift';
import { Button, Form, InputGroup } from 'react-bootstrap';
import MARKETS from '../../markets.json';
import { Dropdown, DropdownContent, DropdownItem } from './styles';

interface Props {
  selectedCountry?: string;
  onChange?: (country: string) => void;
}

/** Market options, extracted from https://simplywall.st/stocks */
export interface Market {
  option: string; // "au",
  label: string; //"Australia",
  href: string; //"/stocks/au/market-cap-large"
}

export const MarketsDropdown: React.FunctionComponent<Props> = ({
  selectedCountry,
  onChange,
}) => {
  {
    const [inputItems, setInputItems] = useState<Market[]>(MARKETS);

    // Pre-select the country passed in props
    const initialSelecteditem = inputItems.find(
      (market) => market.option === selectedCountry
    );

    const itemToString = (item: Market): string => item.label;
    const {
      isOpen,
      getToggleButtonProps,
      getLabelProps,
      getMenuProps,
      getInputProps,
      getComboboxProps,
      highlightedIndex,
      getItemProps,
    } = useCombobox({
      items: inputItems,
      itemToString,
      initialSelectedItem: initialSelecteditem,
      initialInputValue: initialSelecteditem?.label,
      // Pre-select the country passed in props
      onInputValueChange: ({ inputValue }) => {
        setInputItems(
          (MARKETS as Market[]).filter((item) =>
            item.label.toLowerCase().startsWith(inputValue.toLowerCase())
          )
        );
      },
      onSelectedItemChange: ({ selectedItem }) => {
        if (onChange && selectedItem) onChange(selectedItem.option);
      },
    });
    return (
      <Dropdown>
        <Form.Group className="mb-3">
          <div {...getComboboxProps()}>
            <Form.Label htmlFor="country-dropdown" {...getLabelProps()}>
              Choose a market:
            </Form.Label>
            <InputGroup>
              <Form.Control
                id="country-dropdown"
                type="text"
                {...getInputProps()}
              />
              <Button
                variant="outline-secondary"
                id="button-addon2"
                {...getToggleButtonProps()}
                aria-label="toggle menu">
                &#8595;
              </Button>
            </InputGroup>

            <DropdownContent {...getMenuProps()}>
              {isOpen &&
                inputItems.map((item, index) => (
                  <DropdownItem
                    isHighlighted={highlightedIndex === index}
                    key={item.option}
                    {...getItemProps({ item, index })}>
                    {item.label}
                  </DropdownItem>
                ))}
            </DropdownContent>
          </div>
        </Form.Group>
      </Dropdown>
    );
  }
};
