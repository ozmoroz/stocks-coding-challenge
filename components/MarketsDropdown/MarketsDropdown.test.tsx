import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
const { axe } = require('jest-axe');
import { Props, MarketsDropdown } from './MarketsDropdown';

const defaultProps: Props = {
  selectedCountry: 'au',
  onChange: jest.fn(),
};

describe('MarketsDropdown', () => {
  test('shows the pre-selected country', () => {
    const props = { ...defaultProps, selectedCountry: 'be' };
    render(<MarketsDropdown {...props} />);
    expect(screen.getByRole('textbox')).toHaveValue('Belgium');
  });

  test('shows list of countries when "down" combobox button is clicked', () => {
    render(<MarketsDropdown {...defaultProps} />);
    userEvent.click(screen.getByRole('button', { name: 'Open' }));

    expect(screen.getByRole('option', { name: 'Global' })).toBeInTheDocument();
    expect(
      screen.getByRole('option', { name: 'United States' })
    ).toBeInTheDocument();
    expect(
      screen.getByRole('option', { name: 'Australia' })
    ).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Canada' })).toBeInTheDocument();
  });

  test('Choosing a country option fires onChangeEvent', () => {
    render(<MarketsDropdown {...defaultProps} />);
    userEvent.click(screen.getByRole('button', { name: 'Open' }));
    userEvent.click(screen.getByRole('option', { name: 'United States' }));
    expect(defaultProps.onChange).toHaveBeenCalledWith('us');
  });

  test.skip('Pressing escape in the input field clears it', async () => {
    render(<MarketsDropdown {...defaultProps} />);
    const textBox = screen.getByRole('textbox', { name: 'Choose a market:' });
    userEvent.type(textBox, '{esc}');
    // TODO: figure out why this fails
    await waitFor(() => expect(textBox).toHaveValue(''));
  });

  test('Typing into the textbox filter countries', async () => {
    render(<MarketsDropdown {...defaultProps} />);
    const textBox = screen.getByRole('textbox', { name: 'Choose a market:' });
    userEvent.type(textBox, '{esc}'); // Clear the textbox first
    userEvent.type(textBox, 'be');
    expect(
      await screen.findByRole('option', { name: 'Belgium' })
    ).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Bermuda' })).toBeInTheDocument();
    expect(
      screen.queryByRole('option', { name: 'United States' })
    ).not.toBeInTheDocument();
  });

  test('has no accessibility issues', async () => {
    /* Check for basic accessibility issues
     * ref: https://github.com/nickcolley/jest-axe
     */
    const { container } = render(<MarketsDropdown {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
