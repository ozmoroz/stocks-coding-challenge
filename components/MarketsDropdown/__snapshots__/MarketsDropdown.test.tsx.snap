// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketsDropdown matches the snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c1 {
  border-color: rgb(206,212,218);
}

.c2 {
  list-style: none;
  position: absolute;
  background-color: #fff;
  min-width: 160px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
  z-index: 100;
  padding: 0;
}

<div>
  <div
    class="c0"
  >
    <div
      class="mb-3"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="downshift-19-menu"
        role="combobox"
      >
        <label
          class="form-label"
          for="downshift-19-input"
          id="downshift-19-label"
        >
          Choose a market:
        </label>
        <div
          class="input-group"
        >
          <input
            aria-autocomplete="list"
            aria-controls="downshift-19-menu"
            aria-labelledby="downshift-19-label"
            autocomplete="off"
            class="form-control"
            id="downshift-19-input"
            type="text"
            value="Australia"
          />
          <button
            aria-label="toggle menu"
            class="c1 btn btn-outline-secondary"
            id="downshift-19-toggle-button"
            tabindex="-1"
            type="button"
          >
            â†“
          </button>
        </div>
        <ul
          aria-labelledby="downshift-19-label"
          class="c2"
          id="downshift-19-menu"
          role="listbox"
        />
      </div>
    </div>
  </div>
</div>
`;
